### ISSUE TITLE
Support for pretty formatting XML documents

### ISSUE THREAD

**Issue by YorickPeterse** [YorickPeterse/oga#75]
Similar to REXML it would be nice if Oga had a way to spit out a pretty formatted XML document including correct indentation and all that.

## Spec (as of June 27th, 2017)

* `to_xml` on the various node types should take a `**kwargs` argument which is passed to `XML::Generator#initialize`
* `XML::Generator#initialize` should take two additional keyword arguments: `pretty` (default `false`) and `indent` (default 4 spaces), producing the argument list `initialize(root, pretty: false, indent: '    ')`
* A user should be possible to specify its own indent string in `to_xml` using `to_xml(indent: '  ')` to indent with 2 spaces or `to_xml(indent: "\t")` to indent with tabs
* If `pretty` is `true` then XML is pretty printed, and for every element we descend into we add an indentation level. The easiest way is to keep track of an `indent` counter which is passed down into the methods called for generating the XML of elements
* Tests can be added to `spec/oga/xml/generator_spec.rb` in the `#to_xml` `describe` block

**Comment by trevorrowe** [comment]
Did you have an idea of where you would like to hang the pretty formatter? Do you envision this as an extension of `Oga::XML::Document#to_xml`? I would be interested in contributing.

**Comment by YorickPeterse** [comment]
Most likely the actual logic would end up in a separate class called by the `to_xml` methods. I'm not sure yet if I want this to make the default/only behaviour, or if someone would have to call `to_xml` using something like `to_xml(:pretty => true)`. 

One of the tricky things is to get whitespace correct and making sure everything is indented properly, I'm not really sure yet how to handle this.

**Comment by trevorrowe** [comment]
I would expect the default behavior to produce XML without any extra whitespace, preserving only whitespace inside text elements.

I have experience building structured XML documents. I wrote this to be able to remove the `aws-sdk`'s dependency on the `builder` gem: https://github.com/aws/aws-sdk-ruby/blob/version-2.1-refactor/aws-sdk-core/lib/aws-sdk-core/xml/doc_builder.rb

**Comment by YorickPeterse** [comment]
With the new `XML::Generator` class in place this should be easier to implement. To do so basically every  instance of `output <<` in the code would have to take into account any indentation to insert first. The indent level can be increased/decreased in the `to_xml` loop. Basically whenever the loop descends into a child node the indent level would be increased, whenever it moves back to the parent (just after calling `after_element`) it would decrease the level.

Pretty printing XML should be disabled by default.

**Comment by SamSaffron** [comment]
I started something tiny here (not production grade with many glitches, but something) 

https://github.com/discourse/discourse/blob/markdownit/lib/html_normalize.rb 

I think you need a separate normalizer for xml vs html4 vs html5 

Pretty and consistent formatting is super critical for test stability. For example we are switching markdown parsers and the new one generates equivalent html to the old on (semantically) though visually they look different. Figuring out a good strategy for interior spaces is huge. 

In other news I am considering dropping nokogiri from Discourse

**Comment by YorickPeterse** [comment]
@SamSaffron Aha, very interesting! I _think_ we could have 1 formatter for both XML and HTML since the indent rules (from the top of my head). Having said that I myself am super busy, so I'm not sure if I can commit to implementing this myself. However, I'm more than happy to mentor somebody with it and review their code. That way the bus factor also increases from 1 to hopefully 2-ish.

**Comment by SamSaffron** [comment]
@YorickPeterse  I will see how we go, I have a GSOC student I will be tasking at migrating Discourse to oga after that we may be able to look at the formatter, I was thinking of a multiple formatters cause:

```
<p>
   <img src='http://awesome.com/a.png'>
</p>

vs

<p>
   <img src='http://awesome.com/a.png'></img>
</p>
```

btw... quick one, I noticed you decided on mpl 2 license, is there a technical reason you went with that vs the more common mit/bsd? I do not think it creates any problems for us but it does make it harder for lawyers to reason about things. (I do not think we have any MPL 2 code in Discourse at the moment, all 3rd party deps are MIT/BSD)

**Comment by YorickPeterse** [comment]
>  I was thinking of a multiple formatters cause:

I would personally prefer to either indent/format things based on XML, or HTML5; this is basically what Oga does now (sans indentation).

> btw... quick one, I noticed you decided on mpl 2 license, is there a technical reason you went with that vs the more common mit/bsd

Because the MPL is a vastly superiour license to both. It covers more topics (e.g. patents and grants), it's explicit about not granting rights to trademarks. Finally it requires modifications to the code to be made public (when distributed), without going full on crazy like the GPL.

I chose it because I'm fine with people making private modifications they keep to themselves, but I'm not OK with people also distributing those changes (under a different license). I also don't want to deal with nonsense such as contributor agreements for patents as a license should take care of this in my opinion.

So in short, because I found it to be a better license.

**Comment by YorickPeterse** [comment]
>  I will see how we go, I have a GSOC student I will be tasking at migrating Discourse to oga

This sounds really cool! If they need any help, feel free to send them over.

**Comment by timshadel** [comment]
May I request something similar to Nokogiri's API,

`to_xml(indent: 7)`

I use this to format Apple PList files used in iOS apps for translation. I had never seen such an odd number used for indentation before working with those documents, but it'd make transitioning from Nokogiri to Oga remain really easy if the pretty printing options were similar as well. 😄

**Comment by YorickPeterse** [comment]
This has been moved to https://gitlab.com/yorickpeterse/oga/issues/75

**Comment by jmarrec** [comment]
@yorickpeterse I'm a little bit confused. This issue was moved to gitlab, but gitlab repo says 

![Image](https://github.com/user-attachments/assets/b97e385d-6be5-467a-82ec-711f4f9ed18c)

Where is your issue tracker? here or gitlab?

If here, I would kindly ask if you could reopen this one please.
